### ISSUE TITLE
Blocks text placeholder may have insufficient color contrast depending on the active theme

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#60275]
### Description

Some blocks display a placeholder text.
- For blocks that use the RichText component, it's a CSS generated content.
- For blocks that use an input element, it's the HTML placeholder attribute.

In all cases, the placeholder text must meet the color contrast ratio requirement of 4.5:1. It is part of the user interface so that the editor must make sure it is readable for everyone.

Turns out the color of the text placeholder may inherit from the active theme.
For the RichText, the editor CSS sets a CSS `opacity: .62;` regardless.

Setting that opacity doesn't seem to be ideal. When the active theme uses a light color, the opacity set by the editor may bring the contrast ratio below the required threshold. This is the case with Twenty Twenty-Four, for example.

Iin the screenshot below, I'm comparing some block text placeholder in the editor. On the left, with Twenty Twenty-Three active and on the right with Twenty Twenty-Four active.

Observe the color of some of the placeholders actually changes depending on the theme and it has insufficient contrast.

![Twenty Twenty-Three-Four](https://github.com/WordPress/gutenberg/assets/1682452/26905e89-6f9f-4c97-b7a9-da5b467794c8)

### Step-by-step reproduction instructions

- Create a post.
- Add some blocks that have text placeholders, for example:
  - Image (for the caption placeholder)
  - Gallery (for the gallery caption placeholder)
  - Site tagline
  - Site title
  - More 
  - Read more
- Empty any default content from the inputs, to reveal the placeholder text.
- Observe all the placehholder text with different themes active e.g. with Twenty Twenty-Three and with Twenty Twenty-Four.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by afercia** [comment]
Noting that some bocks have some specific handling of the opacity for the placeholder text, which is less than ideal. See for example https://github.com/WordPress/gutenberg/pull/42519 / https://github.com/WordPress/gutenberg/issues/35928

**Comment by afercia** [comment]
Note: there is [a scss mixin](https://github.com/WordPress/gutenberg/blob/4cfaab6302c862ce264247ebef31b45729f04343/packages/base-styles/_mixins.scss#L207-L239) that styles the placeholder attribute but it is only used for some inputs of the editor interface. It's not used for the _blocks_ placeholder.

**Comment by afercia** [comment]
Screenshot with a dark theme (actually a Twenty Twenty-Four style variant):

![Screenshot 2024-03-28 at 11 30 46](https://github.com/WordPress/gutenberg/assets/1682452/29945aae-0fc8-43a3-91c6-acf16b497602)

**Comment by afercia** [comment]
While working on this, I noticed the CSS class `is-dark-theme` is not applied correctly to the _post_ editor. It works as expected in the _site_ editor. That's a blocker for this issue. I will create a separate issue and PR.

**Comment by afercia** [comment]
After some research in the project history, I found some related argumentations in https://github.com/WordPress/gutenberg/pull/33058, specifically:

> we do need the placeholder colors customized for text inputs in the canvas

I don't think the text placeholder _within the editor canvas_ should match the colors set by the theme or customized by the user. The text placeholders aren't part of the site preview. Instead, they are part of the user interface, whether they are placeholder of input elements or the placeholder of the RichText component.

As such, text placeholders in the canvas must always be readable, whatever the background color is. Their color must be set by the editor. Of course, this should not affect text placeholders in other parts of the editor UI, e.g. sidebars, panels, popovers, etc.

One special case that exposes a similar issue is the Cover block. The block detects the color of the image overlay or background color and _dynamically changes_ the text placeholder and also the default cover text color. But, when users set a custom text color, it is set also on the text placeholder. This shouldn't happen because the text placeholder may become unreadable. Instead, an editable input placeholder should always be visible. Animated GIF to illustrate:

![cover](https://github.com/WordPress/gutenberg/assets/1682452/26eec3a1-93af-48da-80e6-0a81406025a1)

**Comment by afercia** [comment]
See related past considerations regarding the missing `is-dark-theme` CSS class on https://github.com/WordPress/gutenberg/issues/64548 / https://github.com/WordPress/gutenberg/pull/64549

**Comment by afercia** [comment]
Note to myself: 
The opacity for the native placeholder attribute e.g. `-webkit-input-placeholder` and the like was originally introduced in https://github.com/WordPress/gutenberg/pull/6406 and then went through various refactorings e.g. in https://github.com/WordPress/gutenberg/pull/9562. Many things have changed since then and now the placeholder mixin seems being used only for the UI controls, not for placeholder attributes in the blocks.
There's separate rulesets for the editor in 'Code editor' mode.

**Comment by himanshupathak95** [comment]
Hey @afercia, 

I have a PR - #69312 that removes the `opacity: 1` CSS property from `::-moz-placeholder` selectors throughout as your issue https://github.com/WordPress/gutenberg/issues/69311 mentions. 

I see your changes here in #69333 also partly remove the `opacity: 1` CSS property at places and create a mixin, which might conflict with #69312. 

So I think we should either ship #69312 first and then proceed with your changes or incorporate all the changes in any one of the PR since the changes are almost similar. This way I think we can avoid conflicts and streamline the approach better. 

What do you think?

**Comment by afercia** [comment]
@himanshupathak95 I would say to merge your https://github.com/WordPress/gutenberg/pull/69312 first as it's simpleer and self-contained. I'll then rebase my PR, no rpoblems.

**Comment by afercia** [comment]
One more finding is about the opacity for the caption placeholder:

https://github.com/WordPress/gutenberg/blob/0d5c3108708266b1cb139087529556027a664a27/packages/block-editor/src/components/rich-text/content.scss#L1-L23

It uses this selector:
`figcaption.block-editor-rich-text__editable [data-rich-text-placeholder]::before`

to increase the opacity for the captions. However, I'm not sure that has ever worked because the placeholder text uses `::after` instead of `::before`. Introduced in https://github.com/WordPress/gutenberg/pull/16733. That PR mentions that at some point the placeholders were first implemented with `::before` so it could be something that was missed at some point.

Also, I would argue that the reasoning that the caption placeholder needed some higher opacity to be better readable, that is valid for all placeholders of all blocks so that at the very least I'd propose to increase the opacity value for the default placeholder to `0.8` or maybe even a little more.

**Comment by carolinan** [comment]
> I don't think the text placeholder within the editor canvas should match the colors set by the theme or customized by the user.

I think it may be confusing for the user if the text changes color when they start typing. But I have no research to back this up with.

A problem that may be specific to the media captions, is that the blocks do not all have their own background and caption text color options. Which means that the color contrast checker is not used where the block is placed in the content.
If the theme colors do not have a high enough contrast, using the theme color for the placeholder is an accessibility issue.

**Comment by afercia** [comment]
> I don't think the text placeholder within the editor canvas should match the colors set by the theme or customized by the user.

I'm reconsidering my previous thinking and at the same time I'm not sure there is an ideal solution. 

One one hand, the RichText placeholder is part of the editing UI. It identifies an editable field. As such, it's sort of the equivalent of the HTML `placeholder` attribute of a native input field. Given it's part of the editing UI, it should always use a color with sufficient contrast ratio. However, hardcoding the color would not guarantee a sufficient contrast because the background can be changed in many ways even by editing the theme's color palette.

On the other hand the RichText placeholder color should be close to the actual text color because it is also sort of a 'preview'. Otherwise, there may be a sudden color change when users start typing, as @carolinan pointed out.

Maybe for now the best option would be to make the RichText placeholder use a color that is close to the block text color i.e. the same color, slighted dimme out.

However, there are cases where the RichText placeholder text is not easily identifiable as a placeholder. For example, there have been user reports about the 'Add read more link` being very confusing. The potential confusion comes from inconsistency in the placeholder usage _and_ from the fact it differs from normal text just by using a slightly different color. There's no other UI or indication to understand the placeholder is... a placeholder.

The screenshot below illustrates some inconsistent usage patterns:
- Add / Write ...: ideally, the placeholder should always either use 'Add' or 'Write' but at least the text clarifies this is a placeholder text.
- 'Read more': in this case the placeholder shows _the default value_ instead of instructions. It's not clear this text is a placeholder.
- 'Add "read more" link text': the placeholder text is underline. It looks like a link, potentially very confusing for users.

![Image](https://github.com/user-attachments/assets/54050e84-e5cc-4535-a64c-f1099c983522)

Some more consistent wording for the placeholders would help.

However, @carolinan suggested, and I would agree, to consider some additional visual indication, maybe an icon or something else, to further clarify the placeholder text is just a placeholder of an editable field.

**Comment by afercia** [comment]
One more finding: the opacity for the Button block placeholder doesn't work as intended because the syntax for the CSS selector is wrong (missing space after the ampersand): 

https://github.com/WordPress/gutenberg/blob/e093fefd041eb6cc4a4e7f67b92ab54fd75c8858/packages/block-library/src/button/editor.scss#L25-L28

This should be removed anyways to inherit from the RichText CSS and avoid exceptions.
